apps-rendering/src/server/insertNewsletterIntoItem.ts

```diff
@@ -0,0 +1,264 @@
+// ----- Imports ----- //
+import type { Newsletter } from '@guardian/apps-rendering-api-models/newsletter';
+import { ArticleDesign } from '@guardian/libs';
+import { OptionKind } from '@guardian/types';
+import type { Body, BodyElement, NewsletterSignUp } from 'bodyElement';
+import { ElementKind } from 'bodyElementKind';
+import type { Item } from 'item';
+import { logger } from 'logger';
+import { Result } from 'result';
+import { stringToPillar } from 'themeStyles';
+
+// ----- Constants ----- //
+
+const enum ElementCategory {
+	'ParagraphText' = 'PARAGRAPH',
+	'BoldParagraphText' = 'BOLD_PARAGRAPH',
+	'NonParagraphText' = 'OTHER TEXT ELEMENT',
+	'Heading' = 'HEADING',
+	'NonText' = 'NON TEXT',
+	'WhiteSpace' = 'WHITE SPACE',
+	'Error' = 'ERROR',
+}
+
+const DEBUG = {
+	log: false,
+	logContent: false,
+};
+
+// ----- pure functions ---//
+
+function categoriseElement(
+	result: Result<string, BodyElement>,
+): ElementCategory {
+	if (!result.isOk()) {
+		// using result.isErr does not convince the compiler that result.value exists
+		return ElementCategory.Error;
+	}
+
+	switch (result.value.kind) {
+		case ElementKind.HeadingThree:
+		case ElementKind.HeadingTwo:
+			return ElementCategory.Heading;
+		case ElementKind.Text: {
+			const { doc } = result.value;
+			if (doc.textContent?.trim().length === 0) {
+				return ElementCategory.WhiteSpace;
+			}
+
+			if (doc.nodeName === 'P') {
+				if (doc.nodeType === 1) {
+					const element = doc as Element;
+					// for paragraphs containing only single <strong> child ("bold text"),
+					// the sign up block can be put before them, but not after them.
+					if (
+						element.children.length === 1 &&
+						element.firstElementChild?.tagName === 'STRONG'
+					) {
+						return ElementCategory.BoldParagraphText;
+					}
+				}
+				return ElementCategory.ParagraphText;
+			} else {
+				return ElementCategory.NonParagraphText;
+			}
+		}
+		default:
+			return ElementCategory.NonText;
+	}
+}
+
+/**
+ * Get the target range of places within a list of visible elements
+ * that the sign-up block can be placed in, being the places within
+ * 3 elements from the middle of the article (by element number only;
+ * not taking account of the length/height of each element)
+ *
+ * @param numberOfElements
+ * @returns The minimum and maximum index for the target range
+ */
+function getRange(numberOfElements: number): [number, number] {
+	const middle = Math.floor(numberOfElements / 2);
+	return [Math.max(0, middle - 3), Math.min(numberOfElements, middle + 3)];
+}
+
+/**
+ *  Find the place to insert a insert a sign-up block in to an article body,
+ *  using these rules:
+ *	 - Don't put straight after bold text
```

Comment:
not quite - `contentBefore` is the element that would be **before the signup block** (so the one the signup block , will go after. 
Now I read that back I can see that's confusing! `contentBefore` could be a name for the content that the signUp will be before or a name for  the content that will be before the signup

So, to look at the whole conditional:
```
			[ElementCategory.ParagraphText].includes(
				categoriseElement(contentBefore),
			) &&
			[
				ElementCategory.ParagraphText,
				ElementCategory.BoldParagraphText,
			].includes(categoriseElement(contentAfter))
```
the place is only suitable if it would go like this:

contentBefore: ParagraphText
[signup goes here]
contentAfter: ParagraphText or BoldParagraphText

Reply:
Thank you for the explanation! That's really helpful. I think I was reading the logic in the wrong direction! The way it is now makes sense